amads/time/meter/__init__.py

```diff
@@ -0,0 +1,619 @@
+"""
+This module serves to map out metrical hierarchies in a number of different ways and
+to express the relationship between
+the notes in
+and
+the hierarchy of
+a metrical cycle.
+
+Uses include identifying notes that traverse metrical levels,
+for analysis (e.g., as a measure of syncopation)
+and notation (e.g., re-notating to reflect the
+within-measure notational conventions).
+"""
+
+from typing import Optional, Union
+
+# ------------------------------------------------------------------------------
+
+
+class MetricalHierarchy:
+    """
+    This class serves to create standard, interoperable representations
+    of metrical hierarchy,
+    centred on a single representation
+    of metrical levels in terms of starts expressed by quarter length
+    from the start of the measure,
+    and (optionally) other representations.
+
+    At least one of the parameters (below) must be specified.
+    These are listed in order of override, e.g., if a full `start_hierarchy` is specified, all else are ignored.
+
+    For a 'quick start', the `time_signature` defaults will serve most use cases, and
+    that along with `levels` should be cover almost all.
+
+    Parameters
+    ----------
+    start_hierarchy
+        Users can specify the `start_hierarchy` directly
+        and completely from scratch (ignoring all other parameters and defaults).
+        Use this for advanced, non-standard metrical structures
+        including those without 2-/3- grouping, or even nested hierarchies.
```

Comment:
Should we state explicitly (and test that) all elements in a given level of the hierarchy should be present in the lower levels of the hierarchy too? In other words, we are assuming that our metrical hierarchy is created by recursively subdividing a unit of time? Maybe worth explicitly stating this, and acknowledging that this means we can't encode polymetres.

Reply:
Yes, as we've discussed offline. I have this on my TODO list for the testing side. Even in cases like 4/4 +3/8, any "levelled out" representation is going to have to have this property. Issue. For now, the comment only.